components/suggest/src/db.rs

```diff
@@ -197,108 +182,185 @@ impl<'a> SuggestDao<'a> {
+                ":keyword": keyword_lowercased,
+                ":provider": SuggestionProvider::Wikipedia
+            },
+            |row| -> Result<Suggestion> {
+                let suggestion_id: i64 = row.get("id")?;
+                let title = row.get("title")?;
+                let raw_url = row.get::<_, String>("url")?;
+
+                let keywords: Vec<String> = self.conn.query_rows_and_then_cached(
+                    "SELECT keyword FROM keywords
+                     WHERE suggestion_id = :suggestion_id AND rank >= :rank
+                     ORDER BY rank ASC",
+                    named_params! {
+                        ":suggestion_id": suggestion_id,
+                        ":rank": row.get::<_, i64>("rank")?,
                     },
-                    SuggestionProvider::Pocket => {
-                        let confidence = row.get("confidence")?;
-                        let full_suffix = row.get::<_, String>("keyword_suffix")?;
-                        let suffixes_match = match confidence {
-                            KeywordConfidence::Low => full_suffix.starts_with(keyword_suffix),
-                            KeywordConfidence::High => full_suffix == keyword_suffix,
-                        };
-                        if suffixes_match {
-
-                                    Ok(Some(Suggestion::Pocket {
-                                        title,
-                                        url: raw_url,
-                                        score,
-                                        is_top_pick: matches!(
-                                            confidence,
-                                            KeywordConfidence::High
-                                        )}))
-                                    } else {
-                            Ok(None)
-                        }
+                    |row| row.get(0),
+                )?;
+                let icon = self.conn.try_query_one(
+                    "SELECT i.data
+                     FROM icons i
+                     JOIN wikipedia_custom_details s ON s.icon_id = i.id
+                     WHERE s.suggestion_id = :suggestion_id",
+                    named_params! {
+                        ":suggestion_id": suggestion_id
                     },
-                    _ => Ok(None),
+                    true,
+                )?;
+                Ok(Suggestion::Wikipedia {
+                    title,
+                    url: raw_url,
+                    full_keyword: full_keyword(keyword_lowercased, &keywords),
+                    icon,
+                })
+            },
+        )?;
+        Ok(suggestions)
+    }
+
+    /// Fetches Suggestions of type Amo provider that match the given query
+    pub fn fetch_amo_suggestions(
+        &self,
+        query: &SuggestionQuery,
+        suggestions_limit: i32,
+    ) -> Result<Vec<Suggestion>> {
+        let keyword_lowercased = &query.keyword.to_lowercase();
```

Comment:
Let's do the `suggestions_limit = query.limit.unwrap_or(-1)` in these functions—`query` already has the limit; it's a little awkward to pass it and the unwrapped limit as arguments.